### PR TITLE
Refactoring of GLPK-based approximating simplex solver #6968

### PR DESCRIPTION
**Pull request by aytey**
This PR implements the changes as discussed in #6968 to clean-up the interface around the `ApproxGLPK` class.

Namely:

- Minimised the interface around `ApproximateSimplex` to expose only what's used by the rest of `cvc5`
- Change most methods in `ApproximateSimplex` to be pure virtual (some methods were `static`, but we actually had a class instance, so there are now instance methods, to allow `ApproximateSimplex` to be an interface class); `mkApproximateSimplexSolver` and `enabled` are left as `static`
- All method bodies and members that were in `ApproximateSimplex` are now in `ApproxGLPK`
- The class `ApproxNoOp` has been removed
- Slight reordering of the code to provide a more logical flow, now that most of the code is under `#ifdef CVC5_USE_GLPK`, and to allow for `ApproxGLPK` to be declared at the top of `approx_simplex.cpp`
- Anything that remains "ordered weirdly" is probably an artefact from renaming `ApproximateSimplex` -> `ApproxGLPK`

Minor bits:

- Ran `git-clang-format`
- Removed `using namespace std` (and added `std::` where required) to ensure consistency in `approx_simplex.cpp`

Signed-off-by: Andrew V. Teylu <andrew@tey.lu>

### CODE REVIEW

**Comment by andrewvaughanj** [comment]
Given GLPK isn't tested via CI, I am not sure what is the best way to ensure that these changes are "okay"; maybe we should wait until #6958 is merged and we have GLPK in the nightlies before merging this?

File: src/theory/arith/approx_simplex.cpp [hunk 1]
```diff
@@ -154,261 +138,49 @@ struct CutScratchPad {
...
-  CutScratchPad d_pad;
+ bool d_solvedRelaxation;
+ bool d_solvedMIP;
 
-  std::vector<Integer> d_denomGuesses;
+ static int s_verbosity;
+
+ CutScratchPad d_pad;
+
+ std::vector<Integer> d_denomGuesses;
```
[the remark this answers is not in the document]
Comment (andrewvaughanj) [reply src/theory/arith/approx_simplex.cpp]: Done!

File: src/theory/arith/approx_simplex.cpp [hunk 2]
```diff
@@ -1785,7 +1740,7 @@ MipResult ApproxGLPK::solveMIP(bool activelyLog){
 //   Node ret = safeConstructNary(nb);
 //   Node rew = Rewriter::rewrite(ret);
 //   if(rew.getNumChildren() < ret.getNumChildren()){
-//     //Debug("approx::") << "explainSet " << ret << " " << rew << endl;
+//     //Debug("approx::") << "explainSet " << ret << " " << rew << std::endl;
 //   }
 //   return rew;
 // }
```
[the remark this answers is not in the document]
Comment (andrewvaughanj) [reply src/theory/arith/approx_simplex.cpp]: Done!

File: src/theory/arith/approx_simplex.cpp [hunk 3]
```diff
@@ -2902,29 +2871,40 @@ bool ApproxGLPK::gaussianElimConstructTableRow(int nid, int M, const PrimitiveVe
...
-      Debug("gaussianElimConstructTableRow") << "selecting " << rid << " " << s << endl;
-      //cout << "selecting " << s << " : complexity " << mult.complexity() << " " << mult << endl;
-      //cout << "selecting " << rid << " " << s << endl;
+      Debug("gaussianElimConstructTableRow")
+          << "selecting " << s << " : " << mult << std::endl;
+      Debug("gaussianElimConstructTableRow")
+          << "selecting " << rid << " " << s << std::endl;
+      // cout << "selecting " << s << " : complexity " << mult.complexity() << "
+      // " << mult << std::endl; cout << "selecting " << rid << " " << s <<
+      // std::endl;
```
[the remark this answers is not in the document]
Comment (andrewvaughanj) [reply src/theory/arith/approx_simplex.cpp]: Done!

File: src/theory/arith/approx_simplex.h [hunk 4]
```diff
@@ -69,24 +69,16 @@ class CutInfo;
 
 class ApproximateSimplex{
  public:
+  /** Is GLPK enabled? */
   static bool enabled();
 
   /**
-   * If glpk is enabled, return a subclass that can do something.
-   * If glpk is disabled, return a subclass that does nothing.
+   * If GLPK is enabled, creates a GPLK-based approxmimating solver.
```
[the remark this answers is not in the document]
Comment (andrewvaughanj) [reply src/theory/arith/approx_simplex.h]: Done!

File: src/theory/arith/approx_simplex.cpp [hunk 5]
```diff
@@ -1017,14 +973,14 @@ ApproximateSimplex::Solution ApproxGLPK::extractSolution(bool mip) const
...
-      CVC5Message() << "assignment " << vi << endl;
+      CVC5Message() << "assignment " << vi << std::endl;
     }
 
     bool useDefaultAssignment = false;
 
     switch(status){
     case GLP_BS:
-      // CVC5Message() << "basic" << endl;
+      // CVC5Message() << "basic" << std::endl;
```
[the remark this answers is not in the document]
Comment (andrewvaughanj) [reply src/theory/arith/approx_simplex.cpp:984]: I'm wondering if I should uncomment these and change them to be `Debug("approx")` prints -- what do you think?

Basically, I think some of these lines could be useful (even though they're currently commented-out).

[the remark this answers is not in the document]
Comment (andrewvaughanj) [reply src/theory/arith/approx_simplex.cpp]: Okay, I did this. We can revert the commit if you prefer to remove this code instead.

**Review by 4tXJ7f — APPROVED** [review]
LGTM, thanks for the updates and sorry about the late review.

**Comment by andrewvaughanj** [comment]
@nafur, @ajreynol what should we do about the debugging information that was removed in https://github.com/cvc5/cvc5/pull/7576? Are you happy for me to reintroduce the bits that I feel shouldn't have been removed (and help with debugging/working with the GLPK integration)?

**Comment by nafur** [comment]
> @nafur, @ajreynol what should we do about the debugging information that was removed in #7576? Are you happy for me to reintroduce the bits that I feel shouldn't have been removed (and help with debugging/working with the GLPK integration)?

Yes, please reintroduce them if you feel they help debugging. My main motivation for removing them was that I wanted to get rid of `Notice()` macro. Just use `Trace` instead.

**Comment by andrewvaughanj** [comment]
> > @nafur, @ajreynol what should we do about the debugging information that was removed in #7576? Are you happy for me to reintroduce the bits that I feel shouldn't have been removed (and help with debugging/working with the GLPK integration)?
> 
> Yes, please reintroduce them if you feel they help debugging. My main motivation for removing them was that I wanted to get rid of `Notice()` macro. Just use `Trace` instead.

Okay, I can (and will) do this 🎉

**Comment by aytey** [comment]
@4tXJ7f: okay, this is now rebased with the debugging information re-introduced (as discussed).

**Comment by aytey** [comment]
🥳

**Pull request merged**